codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/modeled/ModeledShapeValue.java

```diff
@@ -0,0 +1,180 @@
+package software.amazon.polymorph.smithyjava.modeled;
+
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.CodeBlock;
+import com.squareup.javapoet.TypeName;
+import software.amazon.polymorph.smithyjava.generator.Generator;
+import software.amazon.polymorph.smithyjava.generator.library.JavaLibrary;
+import software.amazon.polymorph.smithyjava.nameresolver.Constants;
+import software.amazon.smithy.codegen.core.CodegenException;
+import software.amazon.smithy.model.node.ArrayNode;
+import software.amazon.smithy.model.node.BooleanNode;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.NodeVisitor;
+import software.amazon.smithy.model.node.NullNode;
+import software.amazon.smithy.model.node.NumberNode;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.node.StringNode;
+import software.amazon.smithy.model.shapes.CollectionShape;
+import software.amazon.smithy.model.shapes.DocumentShape;
+import software.amazon.smithy.model.shapes.MapShape;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.UnionShape;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Map;
+
+public class ModeledShapeValue {
+
+    public static CodeBlock shapeValue(JavaLibrary subject, boolean base64BlobStrings, Shape shape, Node value) {
+        return value.accept(new ValueNodeVisitor(subject, base64BlobStrings, shape));
+    }
+
+    // This algorithm is fundamentally a double dispatch,
+    // and could have been a ShapeVisitor instead of a NodeVisitor.
+    // That might have been somewhat cleaner since there are more Shape cases than Node cases,
+    // and the same Node case is used across many shape cases (especially StringNode).
+    // An immutable NodeVisitor is a good sweet spot in terms of performance, however,
+    // since we can use the same visitor across all elements of a List, for example.
+    private static class ValueNodeVisitor implements NodeVisitor<CodeBlock> {
```

Comment:
Praise: I tried to wrap my head around the visitor pattern when I was writing `polymorphy/smithjava`,
but could never managed a good enough understanding to use it.

I am glad we are finally introducing it!